doc/cpp_domdec.rst

```diff
@@ -0,0 +1,186 @@
+describes the cell groups on the local MPI rank.
+
+.. Note::
+    Power users know their model and the hardware resources that
+    they are running on very well, and know best how to partition their model
+    workload to get optimal performance.
+    The :cpp:class:`domain_decomposition` type is kept very simple, and is
+    independent of any load balancing algorithm, so power users can supply their
+    own domain decomposition without using one of the built-in load balancers.
+
+.. cpp:namespace:: arb
+
+.. cpp:function:: domain_decomposition partition_load_balance(const recipe& rec, hw::node_info nd)
+
+    Construct a :cpp:class:`domain_decomposition` that distributes the cells
+    in the model described by :cpp:var:`rec` over the hardware resources described
+    by `hw::node_info`.
+
+Decomposition
+-------------
+
+Documentation for the data structures used to describe domain decompositions.
+
+.. cpp:namespace:: arb
+
+.. cpp:enum-class:: backend_kind
+
+    Used to indicate which hardware backend to use for running a :cpp:class:`cell_group`.
+
+    .. cpp:enumerator:: multicore
+
+        Use multicore backend.
+
+    .. cpp:enumerator:: gpu
+
+        Use gpu back end.
+
+        .. Note::
+            Can only be set if the :cpp:class:`cell_group` type supports the GPU backend.
+
+.. cpp:class:: domain_decomposition
+
+    Describes a domain decomposition. Solely responsible for describing the
+    distribution of cells across cell groups and domains.
+    Holds cell group descriptions (:cpp:member:`groups`) for cells assigned to
+    the local domain, and a helper function (:cpp:member:`gid_domain`) used to
+    look up which domain a cell has been assigned to.
+    Also has meta-data about the number of cells in the global model, and the
+    number of domains over which the model is destributed.
+
+    A load balancing algorithm generates the domain decomposition, for example see
+    :cpp:func:`partition_load_balance`.
+
+    .. cpp:function:: bool is_local_gid(cell_gid_type gid) const
+
+        Tests whether the cell with :cpp:var:`gid` is on the local domain.
+
+    .. cpp:member:: std::function<int(cell_gid_type)> gid_domain
+
+        A callback function that returns the domain id that a cell assigned to
+        (using global identifier :cpp:var:`gid`).
+        Set by the load balancing algorithm that generates the domain decomposition.
+
+    .. cpp:member:: int num_domains
+
+        Number of domains that the model is distributed over.
+
+    .. cpp:member:: int domain_id
+
+        The index of the local domain.
+        Always 0 for non-distributed models, and corresponds to the MPI rank
+        for distributed runs.
+
+    .. cpp:member:: cell_size_type num_local_cells
+
+        Total number of cells in the local domain.
+
+    .. cpp:member:: cell_size_type num_global_cells
+
+        Total number of cells in the global model
+        (sum of :cpp:member:`num_local_cells` over all domains).
```

Comment:
That the sum of the `num_local_cells` must equal `num_global_cells` should be described as an (unenforced) constraint or consistency requirement (there is nothing in principle stopping a user from supplying a domain decomposition object for which this is not true.) We should also state that this sum must (for consistency) equal the number of cells as reported by the recipe.

Reply:
I added a note about these consistency requirements.